ASP.NET Core OData 8 pagination docs (#281)
MicrosoftDocs/OData-docs: Odata-docs/webapi-8/fundamentals/client-driven-paging.md

---
title:  "Client-driven paging in ASP.NET Core OData 8"
description: Client-driven paging in ASP.NET Core OData 8.
date:   2023-01-10
ms.date: 1/10/2023
author: habbes
ms.author: clhabins
---

# Client-driven paging in ASP.NET Core OData 8
**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

> The client can implement paging by telling the OData service how many items to return for a given request. The `$top` query option is used to limit the number of results and the `skip` query option used to specify the offset. For example `GET /Customers?$skip=3$top=5` will return up to 5 items starting from the 4th item in the collection (i.e. after skipping the first 3).

Reviewer: The words "_client can implement paging_" do not do it for me... Especially the word "implement". Maybe "_The client can request the OData service to return a specific number of results_"

Author: What I meant by that is the fact that on the client-side you can have paging UI where the user can select a specific page number, or next/prev pages, and the logic that manages this, to keep track of pages, etc. lies on the client side.

Author: I rephrased it based on the suggestion.

Using these two query options the client can divide a large collection into separate pages and fetch the pages in separate requests. For example, if our collection has 25 items, and we want each page to have 10 items, then we can fetch the different page as follows:
- First page: `$top=10`
- Second page: `$skip=10&$top=10`
- Third page: `$skip=20&top=10` (The last page will return only 5 items)

By default, ASP.NET Core OData 8 limits the maximum value for `$top` to 0. So if you add `$top=2` to your request, you would get an error with a message like:

```json
{
    "error": {
        "code": "",
        "message": "The query specified in the URI is not valid. The limit of '0' for Top query has been exceeded. The value from the incoming request is '2'."
}
```

You can configure the maximum top value using the [`ODataOptions.SetMaxTop()`](/dotnet/api/microsoft.aspnetcore.odata.odataoptions.setmaxtop) when adding OData services to your application. The following snippet would set the maximum allowed `$top` value to 100.

> ```c#

Reviewer: I just noticed that \`\`\`c# renders a little different from \`\`\`csharp

\`\`\`c#:
![image](https://user-images.githubusercontent.com/28291332/214649759-14704bd6-383f-46ff-bee9-09b29b85c059.png)

vs

\`\`\`csharp:
![image](https://user-images.githubusercontent.com/28291332/214649937-23bc7a80-cf19-41e0-9d1b-5fae6e181091.png)

Because of this, I think I'd suggest that you use \`\`\`csharp given that even \`\`\`json is rendered as:
![image](https://user-images.githubusercontent.com/28291332/214650320-d20a82d9-3c26-4204-8f63-001bfe7f8763.png)

Author: You mean the language on the title is uppercase (C#) when using `csharp` and lowercase (c#) when using `c#`? I never noticed that before. Thanks.

```

Now this will support values of up to 100, but return an error if `$top` exceeds 100.

You can also remove the limit by setting it to `null`:

```c#
services.AddOData(options =>
    options.SetMaxTop(null)
    // ...
    );
```

Next, we add the [`EnableQuery`](/dotnet/api/microsoft.aspnetcore.odata.query.enablequeryattribute) attribute to the controller action that handles the request. This attribute automatically applies the OData query options in the request to the results returned by the controller action.

```c#
public class ProductsController : ODataController
{
    [EnableQuery]
    public IQueryable<Product> Get()
    {
        return productsCollection;
    }
}
```

## Benefits of this approach

This client-driven paging approach has a number of benefits:
- It's relatively easy to implement
- It allows the client to fetch pages independently and to jump to an arbitrary page. For example, the client can fetch page 5 without having fetched page 4 previously.
- If the client knows the total size of the collection (e.g. using the `$count` query option), the client can compute the total number of pages as well as generate links for each page in advance.

## Drawbacks

> Paging based on `$top` and `$skip` has a number of drawbacks that may make unsuitable in some scenarios.

Reviewer 2: ```suggestion
Paging based on `$top` and `$skip` has a few drawbacks that may make it unsuitable in some scenarios.
```

    { "Id": 3, "Name": "Product 3" },
    { "Id": 4, "Name": "Product 4" },
    { "Id": 5, "Name": "Product 5" }
]
```

The client can fetch the second page with

```http
GET /Products?$top=2&$skip=2
```

This returns the following result:

```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        { "Id": 4, "Name": "Product 4" },
        { "Id": 5, "Name": "Product 5" }
    ]
}
```

Notice the client did not see product 3 because it was removed before it fetched page 2. In a similar way, if new products were added to the data source before product 2, then the client would see product 2 again in page 2 or later pages.

If the underlying data source collection is not ordered, you may also need to manually order the results to have stable pages, e.g. `GET /Products?$top=2&$skip=2&$orderby=Id`.

Using `$skip` can also lead to performance issues in large datasets in some data source implementation. In many SQL-based database systems `$skip` is implemented using `OFFSET`. A request like `/Products?$skip=50000&$top=100` could be translated in the following SQL query:

```sql
SELECT * FROM products
LIMIT 100, OFFSET 50000;
```

> This query will return 100 rows, from row 50,001 to row 50,100. However it will scan through all the first 50,100 rows in the table, then drop the first 50,000. The larger the offset, the longer the query will take.

Reviewer 2: This explains why `skip` isn't suitable but I was looking (here and in server-side doc) for the explanation of how `skipToken` works internally so that it doesn't suffer the same issues.

Author: See: https://github.com/MicrosoftDocs/OData-docs/pull/281#discussion_r1087443168

This makes `$skip`-based pagination unsuitable for large datasets in many database systems. An alternative approach would be to keep track of where you left off using a unique ordered field. In our example, we can use the `Id` field to keep track of the last product we saw. using this technique, the client can fetch the first page with the following request:

```http
GET /Products?$top=2
```
This will return the following result:

```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        { "Id": 1, "Name": "Product 1" },
        { "Id": 2, "Name": "Product 2" }
    ]
}
```

> When the client gets this ersult, it records that the last product it saw was the product with ID 2. So, to fetch the next page, it makes the following request:

Reviewer 2: ```suggestion
When the client gets this result, it records that the last product it saw was the product with ID 2. So, to fetch the next page, it makes the following request:
```

```http
GET /Products?$filter=Id gt 2&$top=2
```
This fetches only products with id greater than 2.

This roughly translate to the following SQL query:
```sql
SELECT * FROM products
WHERE id > 2
LIMIT 2;
```

> If the database has an ordered index on the `id` column, the this query will scale well and perform significantly better on larger datasets and pages than if it wa using `OFFSET`.

Reviewer 2: ```suggestion
If the database has an ordered index on the `id` column, then this query will scale well and perform significantly better on larger datasets and pages than if it was using `OFFSET`.
```

> This approach also makes the pages more resilient to concurrent changes in the data source. But we lose some features like being able to easily compute the total number pages or to jump to a random page. We are limited to visit the pages sequentially.

Reviewer 2: ```suggestion
This approach also makes the pages more resilient to concurrent changes in the data source. But we lose some features like being able to easily compute the total number of pages or to jump to a random page. We are limited to visiting the pages sequentially.
```

}
```

You can configure the maximum top value using the [`ODataOptions.SetMaxTop()`](/dotnet/api/microsoft.aspnetcore.odata.odataoptions.setmaxtop) when adding OData services to your application. The following snippet would set the maximum allowed `$top` value to 100.

```csharp
services.AddOData(options =>
    options.SetMaxTop(100)
    //...
    );
```

Now this will support values of up to 100, but return an error if `$top` exceeds 100.

You can also remove the limit by setting it to `null`:

```csharp
services.AddOData(options =>
    options.SetMaxTop(null)
    // ...
    );
```

Next, we add the [`EnableQuery`](/dotnet/api/microsoft.aspnetcore.odata.query.enablequeryattribute) attribute to the controller action that handles the request. This attribute automatically applies the OData query options in the request to the results returned by the controller action.

```csharp
public class ProductsController : ODataController
{
    [EnableQuery]
    public IQueryable<Product> Get()
    {
        return productsCollection;
    }
}

> ```

Reviewer: @habbes I don't feel very strongly about this but based on how you have presented this section, I feel that it's a bit hard for someone to know that configuring client-side paging involves 2 steps.
I feel like a reorg of the section could help, starting with a heading level 2 "Configuring client-side paging | Configuring client-side paging in OData".
Then, the next line could be "Configuring client-side paging involves two steps:". Then a numbered list comprising of first, configuring the relevant OData services/features and second about the `EnableQuery` attribute. That way, it'd be clear that the two steps are required.
Some people have a tendency of trying things out as they read and without the indication that it's a two-step process, they could get stuck wondering why it doesn't work after performing step 1 only.

> ## Benefits of this approach

Reviewer: Maybe "Benefits of client-side paging approach" or "Advantages of client-side paging approach"?

This client-driven paging approach has a number of benefits:

> - It's relatively easy to implement

Reviewer 3: ```suggestion
- It's relatively easy to implement.
```

- It allows the client to fetch pages independently and to jump to an arbitrary page. For example, the client can fetch page 5 without having fetched page 4 previously.
- If the client knows the total size of the collection (e.g. using the `$count` query option), the client can compute the total number of pages as well as generate links for each page in advance.

> ## Drawbacks

Reviewer: Maybe "Drawbacks of client-side paging approach" or "Disadvantages of client-side paging approach"

> ## Drawbacks

Reviewer: You used an unordered list in the "Benefits" section but you adopted prose here... Was that because there's only one drawback or because it's easier to talk about it in prose form? Does all the content that follow this section fall under "Drawbacks"? Is there a way you could have broken down the section in an ordered/unordered list?

Author: Good observation. I think opted for a prose format because I explain the drawbacks with detailed examples as opposed to just listing them. I think I mention 2 main drawbacks in this section and the end the section in a way that alludes to server-side paging. I could start by listing the disadvantages, using an ordered list, then follow the list with a detailed explanation for each.

Author: I have re-organized this section to start with a list, then illustrate the issues in more detail. Not sure if that's any better though.

Using these two query options the client can divide a large collection into separate pages and fetch the pages in separate requests. For example, if our collection has 25 items, and we want each page to have 10 items, then we can fetch the different page as follows:

- First page: `$top=10`
- Second page: `$skip=10&$top=10`
- Third page: `$skip=20&top=10` (The last page will return only 5 items)

By default, ASP.NET Core OData 8 limits the maximum value for `$top` to 0. So if you add `$top=2` to your request, you would get an error with a message like:

```json
{
    "error": {
        "code": "",
        "message": "The query specified in the URI is not valid. The limit of '0' for Top query has been exceeded. The value from the incoming request is '2'."
}
```

## Configuring client-side paging in OData

Configuring client-side paging involves two steps:

> 1. Configuring the max `$top` query option value

Reviewer 4: ```suggestion
1. Configuring the `$top` query option with a max value
```

2. Adding `EnableQuery` attribute to the controller method

You can configure the maximum top value using the [`ODataOptions.SetMaxTop()`](/dotnet/api/microsoft.aspnetcore.odata.odataoptions.setmaxtop) when adding OData services to your application. The following snippet would set the maximum allowed `$top` value to 100.

```csharp
services.AddOData(options =>
    options.SetMaxTop(100)
    //...
    );
```

Now this will support values of up to 100, but return an error if `$top` exceeds 100.

> You can also remove the limit by setting it to `null`:

Reviewer 4: does this mean when you set it to null you can use $top with whichever value and it will work.. I think you should say that explicitly if that's the case

Author: It means there's no (upper) limit. Any `$top` value >= 0 will be allowed.
